src/t/d15.clj

```diff
@@ -0,0 +1,298 @@
+(ns t.d15
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]
+    [clojure.core.match :refer [match]]))
+
+(def sample1
+"########
+#..O.O.#
+##@.O..#
+#...O..#
+#.#.O..#
+#...O..#
+#......#
+########
+
+<^^>>>vv<v>>v<<")
+
+(def sol1 2028)
+
+(def sample2
+"##########
+#..O..O.O#
+#......O.#
+#.OO..O.O#
+#..O@..O.#
+#O#..O...#
+#O..O..O.#
+#.OO.O.OO#
+#....O...#
+##########
+
+<vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
+vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
+><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
+<<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
+^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
+^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
+>^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
+<><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
+^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
+v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^")
+(def sol2 10092)
+(def sol2p2 9021)
+
+(def sample3
+"#######
+#...#.#
+#.....#
+#..OO@#
+#..O..#
+#.....#
+#######
+
+<vv<<^^<<^^")
+(def sol3p2 (+ 105 2 700 3 600))
+
+(defn parse-input
+  [input]
+  (let [[warehouse moves] (cljstr/split input #"\n\n")
+        warehouse (->> warehouse
+                       cljstr/split-lines
+                       (map-indexed (fn[y l]
+                                      (apply merge-with into
+                                      (map-indexed
+                                        (fn [x c] (case c
+                                                    \# {:walls #{[x y]}}
+                                                    \O {:box #{[x y]}}
+                                                    \@ {:robot #{[x y]}}
+                                                    {}))
+                                        l))))
+                       (apply merge-with into))
+        moves (-> moves
+                  (cljstr/replace "\n" ""))]
+  [warehouse moves]))
+
+(defn symb-to-dir
+  [symb]
+  (case symb
+    \> [+1 0]
+    \< [-1 0]
+    \v [0 +1]
+    \^ [0 -1]))
+
+(defn robot-move-box
+  [robot box1 dir box walls]
+  (loop [might-move #{}
+         try-move box1
+         boxes box]
+    (let [npos (map + try-move dir)]
+      (cond
+        (contains? boxes npos)
+        (recur (cljset/union might-move #{try-move}) npos (disj boxes npos))
```

Comment:
Why not `conj`? Am I missing how this is different?
```suggestion
        (recur (conj might-move try-move) npos (disj boxes npos))
```

Reply:
Functions in clojure.core are, broadly speaking, either "collection" functions or seq ("sequence") functions. Collection functions preserve their input type, whereas seq functions always convert their input to a seq first (typically by calling `seq`) and thus "lose" the type information.

As a rough list off the top of my head:
- `conj`, `disj`, `assoc`, `dissoc`, `count`, `peek`, `pop` are "collection" funcions.
- `first`, `rest`, `last`, `butlast`, `next`, `map`, `filter`, `reduce`, `remove`, `group-by`, `sort` are sequence functions.

It's a very important distinction that may not be all that apparent if you've not been told about it. I don't think there's a way around memorizing which category a function belongs to as part of learning that function.

Reply:
Thanks, I guess I will keep testing until I know it. But it feels like there is a sense underneath all